packages/proxy/lib/http/util/cookies.ts

```diff
@@ -0,0 +1,166 @@
+import _ from 'lodash'
+import type Debug from 'debug'
+import { URL } from 'url'
+import { cors } from '@packages/network'
+import { Cookie, CookieJar } from '@packages/server/lib/cookie-jar'
+
+interface AutomationCookie {
+  domain: string
+  expiry: number | null
+  httpOnly: boolean
+  maxAge: number | null
+  name: string
+  path: string | null
+  sameSite: string
+  secure: boolean
+  value: string
+}
+
+interface RequestDetails {
+  url: string
+  isAUTFrame: boolean
+  needsCrossOriginHandling: boolean
+}
+
+// Sets SameSite context to match what it would be in the browser
+// see https://github.com/salesforce/tough-cookie#samesite-cookies
+export const getSameSiteContext = (autUrl: string | undefined, requestUrl: string, isAUTFrameRequest: boolean) => {
+  // if there's no AUT URL, it's a request for the first URL visited,
+  // so there would be no cross-origin concerns
+  // if the request origin matches the AUT origin, cookies can be handled
+  // in a strict fashion
+  if (!autUrl || cors.urlOriginsMatch(autUrl, requestUrl)) {
+    return 'strict'
+  }
+
+  // being an AUT frame request means it's from navigation, so the context is
+  // 'lax'. otherwise, 'none' indicates a non-navigation cross-origin request
+  return isAUTFrameRequest ? 'lax' : 'none'
+}
+
+const sameSiteNoneRe = /; +samesite=(?:'none'|"none"|none)/i
+
+export const parseCookie = (cookie) => {
+  const toughCookie = CookieJar.parse(cookie)
+
+  if (!toughCookie) return
+
+  // fixes tough-cookie defaulting undefined/invalid SameSite to 'none'
+  // https://github.com/salesforce/tough-cookie/issues/191
+  const hasUnspecifiedSameSite = toughCookie.sameSite === 'none' && !sameSiteNoneRe.test(cookie)
+
+  // not all browsers currently default to lax, but they're heading in that
+  // direction since it's now the standard, so this is more future-proof
+  if (hasUnspecifiedSameSite) {
+    toughCookie.sameSite = 'lax'
+  }
+
+  return toughCookie
+}
+
+const comparableCookieString = (toughCookie) => {
+  return _(toughCookie)
+  .pick('key', 'value', 'domain', 'path')
+  .toPairs()
+  .sortBy(([key]) => key)
+  .map(([key, value]) => `${key}=${value}`)
+  .join('; ')
+}
+
+const areCookiesEqual = (cookieA, cookieB) => {
+  return comparableCookieString(cookieA) === comparableCookieString(cookieB)
+}
+
+const matchesPreviousCookie = (previousCookies, cookie) => {
```

Comment:
```suggestion
const matchesPreviousCookie = (previousCookies: Cookie[], cookie: Cookie) => {
```

Reply:
Addressed in [`95124a3` (#22320)](https://github.com/cypress-io/cypress/pull/22320/commits/95124a38297f4924a8a11dec21668171fa50d514)